### DISCUSSION TITLE
removing changing / document assembly restrictions using modify=all

### DISCUSSION THREAD

**Discussion by miraklemax** [qpdf/qpdf#880] [Q&A]
new to QPDF. I downloaded it and run this command on a PDF file that came from HelloSign (the PDF does not allow rearranging pages or deleting pages but is NOT password protected):

"C:\Program Files\qpdf 11.2.0\bin\qpdf.exe" --encrypt "" ""  256 --modify=all  --  \\EAA-DCFS03\Users\ckubica\Desktop\in.pdf \\EAA-DCFS03\Users\ckubica\Desktop\out.pdf

qpdf creates the out.pdf without error, but the resulting file has Document Assembly and Changing still marked as "Not Allowed."

anyone know why? I thought --modify=all would turn all that back on in the PDF.

Thanks!

**Comment by jberkenbilt** [comment]
When I try this, I don't see what you're seeing, but there are some known issues (hopefully fixed in the next release) around qpdf not removing restrictions for digitally signed files. That said, rather than re-encrypting, you could use the `--decrypt` flag. If all goes well, I'll add something like `--remove-restrictions` (TBD) in the next release which will be 11.3.0.

**Reply by jberkenbilt** [reply]
In other words `qpdf --decrypt in.pdf out.pdf`

**Reply by miraklemax** [reply]
thanks for the suggestions! I did try decrypt at first and even remove restrictions, but it seemed to have no effect at all on the pdf. the in and out files were identical. anyway, if it helps troubleshoot, i was using a digitally signed document generated by HelloSign (the docusign competitor owned by dropbox). I will look for version 11.3.0. When will that be out? ;-) thanks for your work on this app.

**Reply by jberkenbilt** [reply]
I'll try to get it out in the next couple of weeks. Take a look at #833 for a discussion of this problem. The workaround there is a bit more advanced, but that's the behavior I intend to work in for the upcoming release. In case it's not obvious, any of these operations invalidate the signature.

**Reply by miraklemax** [reply]
thank you, sir. if you can work this in, it would be a huge win for my organization. If so, I would like to donate $ if you have a donate page. Yes, I know it invalidate signatures...this is mostly so we can combine signature docs with other brief packages when we can only submit one evidentiary PDF to certain courts. Thanks, again! Look forward to the next version.

**Reply by jberkenbilt** [reply]
I just pushed a change that adds the `--remove-restrictions` command-line flag. This will be in 11.3.0. I don't have a donate page. People have sometimes donated to me personally via PayPal for work on qpdf, but I don't do this for financial donations. My work on qpdf is my gift to the community. As someone who has spent 30+ years working as a professional software developer and benefiting greatly from high quality open source software, it is my honor to be able to give something back, and I am not dependent on income from qpdf. Anyway, actually adding this feature only took about half an hour. :-)

**Reply by jberkenbilt** [reply]
I expect to release 11.3.0 in the next two weeks or so, but that is not a hard commitment.

**Reply by miraklemax** [reply]
thanks. i will look for the new version so I can try out the new functionality on my docs!

**Reply by miraklemax** [reply]
hi! I just downloaded and tried 11.3 but unfortunately, the new version has no effect on my PDFs. The input pdf file is locked from editing by HelloSign and the resulting output PDF is still not changeable. I thought that was the point of the new release, unless I'm doing it wrong. My command is:

I:\>\\EAA-DCFS03\Users\ckubica\Desktop\qpdf-11.3.0-msvc64\bin\qpdf.exe \\EAA-DCFS03\Users\ckubica\Desktop\locked.pdf --remove-restrictions \\EAA-DCFS03\Users\ckubica\Desktop\out.pdf

**Reply by jberkenbilt** [reply]
You need `--decrypt` and `--remove-restrictions`. Those are two different operations, and sometimes you might want one without the other. Does it work if you do both options?

**Comment by miraklemax** [comment]
I tried it with —decrypt as well and it made no difference, alas. Have you tested with a document created and locked by hellosign specifically?

Chris Kubica
President, Founder
Application Architects

1-888-896-4608
www.applicationarch.com

On Mar 27, 2023, at 5:36 PM, Jay Berkenbilt ***@***.***> wrote:

﻿


You need --decrypt and --remove-restrictions. Those are two different operations, and sometimes you might want one without the other. Does it work if you do both options?

**Reply by miraklemax** [reply]
if it helps in testing, here is a test document with pretend data in it that you can try to unlock yourself: https://www.dropbox.com/s/345j7ulv4d8oih7/2023_Renewal_Package_for_19105_Crawford_Ave.__Store_393__16023-11_.pdf?dl=0

**Reply by miraklemax** [reply]
?

**Reply by jberkenbilt** [reply]
I have it starred in my qpdf inbox. I will take a look when I have some time, hopefully this week.

**Comment by miraklemax** [comment]
i thank you kindly! 

Sincerely,
 
Chris Kubica
President, Founder
Application Architects

***@***.***
http://www.applicationarch.com

1-919-259-8023 (iPhone)
1-888-896-4608 (Ph/Fax)
1-202-280-2042 (Int'l Ph/Fax)





On Apr 3, 2023, at 10:38 AM, Jay Berkenbilt ***@***.***> wrote:


I have it starred in my qpdf inbox. I will take a look when I have some time, hopefully this week.

**Reply by miraklemax** [reply]
any word on an update? thanks!

**Reply by jberkenbilt** [reply]
Sorry, I've been slammed. I haven't forgotten, but I can't commit to a particular time frame.

**Reply by jberkenbilt** [reply]
It is near the top of my list.

**Reply by miraklemax** [reply]
thanks!! :-)

**Comment by jberkenbilt** [comment]
I finally got a chance to look at this, but unfortunately when I run `qpdf --remove-restrictions` on the file that you linked to, I can edit it without any issues in Adobe Acrobat Pro.

**Comment by jberkenbilt** [comment]
This is the exact command I used.
```
qpdf --remove-restrictions ~/Downloads/2023_Renewal_Package_for_19105_Crawford_Ave.__Store_393__16023-11_.pdf /tmp/1.pdf
```
After this, I can upload /tmp/1.pdf to Acrobat online and edit it. I cannot edit the original file.

**Comment by miraklemax** [comment]
Ok. Let me try it. Which version of qpdf did you use, just to be sure?


Chris Kubica
President, Founder
Application Architects

1-888-896-4608
www.applicationarch.com

On May 6, 2023, at 10:46 AM, Jay Berkenbilt ***@***.***> wrote:

﻿


This is the exact command I used.


qpdf --remove-restrictions ~/Downloads/2023_Renewal_Package_for_19105_Crawford_Ave.__Store_393__16023-11_.pdf /tmp/1.pdf


After this, I can upload /tmp/1.pdf to Acrobat online and edit it. I cannot edit the original file.

**Reply by jberkenbilt** [reply]
I ran it directly from the source tree, but there have been no changes since 11.3.0 in any way related to this.

**Reply by miraklemax** [reply]
hi! I can confirm that it seems to work as advertised in Adobe Acrobat. So that's great. Did you try to delete a page, though? If I try to do that, I get this error:
<img width="942" alt="Screenshot 2023-05-16 at 11 04 06 AM" src="https://github.com/qpdf/qpdf/assets/123600964/86a38801-9913-4581-9da1-30f51d7e32c8">

**Comment by jberkenbilt** [comment]
I haven't tried to do that. I suppose my code doesn't remove the signature, though it does invalidate it. You can use qpdf to remove a page from the document. qpdf certainly doesn't prevent you from doing that. I suppose it would be possible to add a `--remove-signatures` option as an enhancement, though I don't think it would be high priority.
